src/python/turicreate/toolkits/_image_feature_extractor.py

```diff
@@ -82,83 +82,108 @@ def __init__(self, ptModel):
         self.model = keras.models.load_model(model_path)
 
     def extract_features(self, dataset, feature, batch_size=64, verbose=False):
-        """
-        Parameters
-        ----------
-        dataset: SFrame
-            SFrame of images
-        """
-        from ._mxnet._mx_sframe_iter import SFrameImageIter as _SFrameImageIter
         from six.moves.queue import Queue as _Queue
         from threading import Thread as _Thread
+        import numpy as _np
         import turicreate as _tc
         import array
 
         if len(dataset) == 0:
             return _tc.SArray([], array.array)
 
-        batch_size = min(len(dataset), batch_size)
-        # Make a data iterator
-        dataIter = _SFrameImageIter(sframe=dataset, data_field=[feature], batch_size=batch_size, image_shape=self.image_shape)
+        # Only expose the feature column to the SFrame-to-NumPy loader.
+        image_sf = _tc.SFrame({'image' : dataset[feature]})
+
+        # Encapsulate state in a dict to sidestep variable scoping issues.
+        state = {}
+        state['num_started'] = 0       # Images read from the SFrame
+        state['num_processed'] = 0     # Images processed by TensorFlow
+        state['total'] = len(dataset)  # Images in the dataset
 
-        # Setup the MXNet model
-        model = MXFeatureExtractor._get_mx_module(self.ptModel.mxmodel,
-                self.data_layer, self.feature_layer, self.context, self.image_shape, batch_size)
+        # We should be using SArrayBuilder, but it doesn't accept ndarray yet.
+        #out = _tc.SArrayBuilder(dtype = array.array)
+        state['out'] = _tc.SArray(dtype=array.array)
 
-        out = _tc.SArrayBuilder(dtype = array.array)
-        progress = { 'num_processed' : 0, 'total' : len(dataset) }
         if verbose:
             print("Performing feature extraction on resized images...")
 
-        # Encapsulates the work done by the MXNet model for a single batch
+        # Provide an iterator-like interface around the SFrame.
+        def has_next_batch():
+            return state['num_started'] < state['total']
+
+        # Yield a numpy array representing one batch of images.
+        def next_batch():
+            # Compute the range of the SFrame to yield.
+            start_index = state['num_started']
+            end_index = min(start_index + batch_size, state['total'])
+            state['num_started'] = end_index
+
+            # Allocate a numpy array with the desired shape.
+            # TODO: Recycle allocations?
```

Comment:
I'm not certain I understand this TODO. You mean reuse arrays returned by`np.zeros`?

Reply:
Yup, that's right. I clarified the comment. Hopefully only a performance bottleneck on CPU-only cases anyway